lib/spack/spack/test/compilers.py

```diff
@@ -227,3 +242,14 @@ def test_xl_r_flags():
     unsupported_flag_test("cxx11_flag", "xl_r@13.0")
     supported_flag_test("cxx11_flag", "-qlanglvl=extended0x", "xl_r@13.1")
     supported_flag_test("pic_flag", "-qpic", "xl_r@1.0")
+
+
+@pytest.mark.regression('10191')
+def test_clang_version_detection(echo_executable_string):
+    version = spack.compilers.clang.Clang.default_version(
+        'clang version 6.0.1-svn334776-1~exp1~20181018152737.116 (branches/release_60)\n'  # noqa
+        'Target: x86_64-pc-linux-gnu\n'
+        'Thread model: posix\n'
+        'InstalledDir: /usr/bin\n'
+    )
+    assert version == '6.0.1'
```

Comment:
I've been meaning to add a test like this. We should really add this for every compiler (and several versions of each compiler on multiple OSes). I think most of the compiler definitions have a sample version string you can use. It doesn't have to be in this PR obviously.

Reply:
I extended the test to all the other outputs of Clang that are reported in the docstring. As soon as this PR is merged I'll submit another to cover for the other supported compilers.